src/tcp_ip/flow.cpp

```diff
@@ -110,24 +110,27 @@ void Flow::process_packet(PDU& pdu) {
     }
     const uint32_t chunk_end = tcp->seq() + raw->payload_size();
     const uint32_t current_seq = data_tracker_.sequence_number();
-    // If the end of the chunk ends after our current sequence number, process it.
-    if (seq_compare(chunk_end, current_seq) >= 0) {
-        uint32_t seq = tcp->seq();
-        // If we're going to buffer this and we have a buffering callback, execute it
-        if (seq > current_seq && on_out_of_order_callback_) {
-            on_out_of_order_callback_(*this, seq, raw->payload());
-        }
-        if (data_tracker_.process_payload(seq, move(raw->payload()))) {
-            if (on_data_callback_) {
-                on_data_callback_(*this);
-            }
+    // If the end of the chunk ends before the current sequence number or
+    // if we're going to buffer this and we have a buffering callback, execute it
+    if (seq_compare(chunk_end, current_seq) < 0 ||
```

Comment:
By reordering I did not (hopefully) change the semantics of this part of the code. Note that the OOO callback was also called if "seq_compare(chunk_end, current_seq) < 0". I would not actually mind if we would offer a different callback for duplicates, or at least provide this information in a boolean parameter to the callback (since as it is, my callback has to make this distinction by hand). Dropping this callback might not be a good idea, since we are just adding the ability to skip ahead, so this data might no longer be "duplicate" :)

I reordered the code to combine the two cases where the OOO callback was called and make sure that after the callback the normal "data_tracker_.process_payload" is called, if applicable. The old code could call an OOO callback (from the else part) and NOT call "data_tracker_.process_payload" afterwards (which was irrelevant without the ability to skip ahead in the flow).